Bioinf_tools/bioinf.py

```diff
@@ -0,0 +1,124 @@
+from typing import Union, List, Set
+
+from modules import write_fastq, utils
+from modules.utils import is_bounded
+
+operations = {
+    'transcribe': utils.transcribe,
+    'reverse': utils.reverse,
+    'complement': utils.complement,
+    'transcribe_dna_complement': utils.transcribe_dna_complement,
+    'reverse_complement': utils.reverse_complement
+}
+
+
+def filter_fastq(input_fastq: str, output_fastq: str, gc_bounds: Union[int, tuple[int, int]] = None,
+                 length_bounds: Union[int, tuple[int, int]] = None, quality_threshold: int = None) -> None:
+    """
+    Function filter_fastq
+    Application - The function input is the name of the file with data in fastq format. Using the parameters gc_bounds, length_bounds, and
+    the quality_threshold function filters the lines of the file and writes to the final file only those that match
+    given conditions. To write a file, the function uses the write_fastq module from modules. Writing to a file occurs line by line.
+    If the file already exists, it will not be overwritten and an error will be displayed.
+    Attributes
+    ----------
+    input_fastq: str
+        Input file name
+    output_fastq: str
+        Output file name
+    gc_bounds: int или tuple[int,int]
+        gc_bounds - интервал GC состава (в процентах).
+        Если в аргумент передать одно число, то считается, что это верхняя граница.
+    length_bounds: int или tuple[int,int]
+        Интервал длины для фильтрации. Если в аргумент передать одно число, то считается, что это верхняя граница.
+    quality_threshold: int
+        Пороговое значение среднего качества рида для фильтрации, по-умолчанию равно 0 (шкала phred33).
+    """
+    if length_bounds is None:
+        length_bounds = (0, 2 ** 32)
+    if gc_bounds is None:
+        gc_bounds = (0, 100)
+    if quality_threshold is None:
+        quality_threshold = 0
+    if isinstance(gc_bounds, int):
+        gc_bounds = (0, gc_bounds)
+    if isinstance(length_bounds, int):
+        length_bounds = (0, length_bounds)
+    try:
+        file = open('filtered/' + output_fastq)
+    except IOError:
+        pass
+    else:
+        print('Файл уже существует')
+        return
+    with open(input_fastq) as file:
+        seq_name: str = file.readline().split()[0]
+        seq: str = file.readline().rstrip('\n')
+        quality_name: str = file.readline().rstrip('\n')
+        quality: str = file.readline().rstrip('\n')
```

Comment:
Как предложение - вынести чтение fastq файла в отдельную функцию, чтобы потом сделать чтение и запись на ходу

```suggestion
    def read_fastq_record(file) -> Union[None, Tuple[str, str, str, str]]:
        seq_name = file.readline().strip()
        if not seq_name:
            return None  # End of file
        seq = file.readline().strip()
        quality_name = file.readline().strip()
        quality = file.readline().strip()
        return seq_name, seq, quality_name, quality
```

``` python
with open(input_fastq) as infile, open('filtered/' + output_fastq, 'w') as outfile:
        while True:
            record = read_fastq_record(infile)
            if record is None:
                break  # End of file

            seq_name, seq, quality_name, quality = record
            
```